### PR TITLE
libs/libc/crc16: Separate implementation of crc16xmodem from crc16

### PR DESCRIPTION
**Pull request by anchao**
## Summary

libs/libc/crc16: Separate implementation of crc16xmodem from crc16

keep default crc16 catalogue for CRC-16/XMODEM

Mapping crc16 implement to crc16xmodem

crc16     -> crc16xmodem
crc16part -> crc16xmodempart

- CRC-16/ACORN, CRC-16/LTE, CRC-16/V-41-MSB, XMODEM, ZMODEM
poly: 0x1021 initial seed: 0x0000, xor output: 0x0000

: width=16
: poly=0x1021
: init=0x0000
: refin=false
: refout=false
: xorout=0x0000
: check=0x31c3
: residue=0x0000
: name="CRC-16/XMODEM"

https://reveng.sourceforge.io/crc-catalogue/16.htm#crc.cat.crc-16-xmodem

Signed-off-by: chao an <anchao.archer@bytedance.com>

## Impact

N/A

## Testing

crc16 test via rpmsg uart

### CODE REVIEW

**Comment by nuttxpr** [comment]
[**\[Experimental Bot, please feedback here\]**](https://github.com/search?q=repo%3Aapache%2Fnuttx+13552&type=issues)



Yes, this PR mostly meets the NuttX requirements, but could be improved.

**Strengths:**

* **Clear Summary:** The summary explains the *why*, *what*, and *how* of the change, referencing relevant external resources (Linux kernel implementation and CRC catalogues).  The reasoning for changing the default is well-explained. Renaming the old functions to `crc16xmodem` is also clearly stated.
* **Impact: ** Acknowledges dependency on a `nuttx-apps` PR. This is important for reviewers.
* **Testing:**  Indicates *how* testing was performed ("crc16 test via rpmsg uart").


**Weaknesses and Suggestions for Improvement:**

* **Impact Section Incomplete:**  The impact section needs to be fully filled out.  While referencing the  `nuttx-apps` PR is good, the impact on the core NuttX OS itself needs to be assessed directly in this PR. Consider the following:
    * **Is new feature added?**:  Arguably, CRC-16/IBM is a new feature, even though another CRC existed before.
    * **Is existing feature changed?**: Yes, the default CRC is changed, and the old CRC functions are renamed.  This is a significant change.
    * **Impact on user:** YES. Users relying on the default `crc16` behavior *will* need to adapt. They'll need to switch to `crc16xmodem` or explicitly configure CRC-16/IBM if that was their intended behavior. This needs to be clearly explained.
    * **Impact on build:**  Probably NO.  Clarify this.
    * **Impact on hardware:**  Likely NO.  Clarify this.
    * **Impact on documentation:** YES.  Documentation needs to be updated to reflect the new default and the renamed functions. This PR should ideally include those documentation changes.
    * **Impact on security:** Potentially YES.  If CRC16 was used for any security-sensitive operations (though unlikely), changing the default algorithm could have security implications. Consider this carefully and document the findings.
    * **Impact on compatibility:** YES.  Backward compatibility is broken for code that implicitly relied on the previous default CRC.  Clearly document the migration path.
* **Missing Test Logs:** The "Testing logs before change" and "Testing logs after change" sections are empty. Include actual logs demonstrating the change in behavior.  Show the output of the CRC calculations before and after, confirming the expected values.
* **Consider a Kconfig Option:** Since this is a breaking change, consider adding a Kconfig option to allow users to select the default CRC algorithm at compile time.  This would preserve backward compatibility for existing users while providing the new default.



By addressing these weaknesses, the PR will be much stronger and easier for reviewers to assess.

File: libs/libc/misc/lib_crc16.c [hunk 1]
```diff
@@ -18,107 +18,49 @@
...
-  size_t i;
-
-  for (i = 0; i < len; i++)
-    {
-      crc16val = crc16_tab[((crc16val >> 8) & 0xff) ^
-                 src[i]] ^ (crc16val << 8);
-    }
-
-  return crc16val;
+  return crc16ibmpart(src, len, crc16val);
```
Comment (xiaoxiang781216) [inline libs/libc/misc/lib_crc16.c:49]: it's a breaking change.

Comment (anchao) [reply libs/libc/misc/lib_crc16.c:49]: image
Answer from yourself

Comment (xiaoxiang781216) [reply libs/libc/misc/lib_crc16.c:49]: but, the community has the different opinion, you should post to dev@ list to vote.

**Comment by lupyuen** [comment]
We'll change "Draft Mode" to "Ready For Review" after the voting ends in 72 hours. Thanks :-)

**Comment by anchao** [comment]
> We'll change "Draft Mode" to "Ready For Review" after the voting ends in 72 hours. Thanks :-)

@lupyuen Sorry for bother you again. I will keep the default crc16 algorithm at crc16-xmodem, so this PR is not a breaking change and we can switch to review status.

**Comment by simbit18** [comment]
Hi @anchao 
I think it is also necessary to update the Cmakefile

```
FAILED: nuttx 
: && /tools/ccache/bin/arm-none-eabi-g++  --specs=nosys.specs   -Wl,--print-memory-usage -Wl,--entry=__start -nostdlib -Wl,--gc-sections -Wl,--cref -Wl,-Map=nuttx.map @CMakeFiles/nuttx.rsp -o nuttx  && :
/tools/gcc-arm-none-eabi/bin/../lib/gcc/arm-none-eabi/13.2.1/../../../../arm-none-eabi/bin/ld: warning: nuttx has a LOAD segment with RWX permissions
/tools/gcc-arm-none-eabi/bin/../lib/gcc/arm-none-eabi/13.2.1/../../../../arm-none-eabi/bin/ld: libs/libc/libc.a(lib_crc16.c.obj): in function `crc16part':
/github/workspace/sources/nuttx/libs/libc/misc/lib_crc16.c:49:(.text.crc16part+0x24): undefined reference to `crc16xmodempart'
/tools/gcc-arm-none-eabi/bin/../lib/gcc/arm-none-eabi/13.2.1/../../../../arm-none-eabi/bin/ld: libs/libc/libc.a(lib_crc16.c.obj): in function `crc16full':
/github/workspace/sources/nuttx/libs/libc/misc/lib_crc16.c:65:(.text.crc16full+0x26): undefined reference to `crc16xmodempart'
```

https://github.com/apache/nuttx/actions/runs/14353366426/job/40237220865?pr=16162#step:7:1219

**Comment by anchao** [comment]
> Contributor

@simbit18 Thank you, I updated the cmake, please review again

**Review by xiaoxiang781216 — APPROVED** [review]

**Review by raiden00pl — APPROVED** [review]

**Review by simbit18 — APPROVED** [review]

**Review by jerpelea — APPROVED** [review]
